File: examples/js/expo/gnoboard/android/app/src/main/java/land/gno/gobridge/JavaPromiseBlock.java
package land.gno.gobridge;

import com.facebook.react.bridge.Promise;
import gnolang.gno.gnonative.PromiseBlock;
import java.util.ArrayList;
import java.util.List;

class JavaPromiseBlock implements PromiseBlock {
    public static List<JavaPromiseBlock> promises = new ArrayList<>();
    private final Promise promise;

    public JavaPromiseBlock(Promise promise) {
        this.promise = promise;
        this.store();
    }
    public void callResolve(String reply) {
        this.promise.resolve(reply);
        this.remove();
    }

    public void callReject(java.lang.Exception error) {
        this.promise.reject(error);
        this.remove();
    }

    private void remove() {
        promises.remove(this);
    }

    private void store() {
        promises.add(this);
    }
}

File: expo/android/src/main/java/land/gno/gnonative/JavaPromiseBlock.kt
package land.gno.gnonative

//  Created by Guilhem Fanton on 10/07/2023.

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import gnolang.gno.gnonative.PromiseBlock as IPromiseBlock

class PromiseBlock(val promise: Promise): IPromiseBlock {
  // gnolang.gno.gnonative.PromiseBlock aims to keep reference over promise object so go can play with
  // until the promise is resolved
  companion object {
    private var promises = mutableSetOf<PromiseBlock>()
  }

  init {
    store()
  }

  override fun callResolve(reply: String?) {
    this.promise.resolve(reply ?: "")
    this.remove() // cleanup the promise
  }

  override fun callReject(err: Exception?) {
    if (err?.message == "EOF") {
      this.promise.reject(GoBridgeCoreEOF())
    } else {
      // Only the reject()'s message argument will be thrown to React Native, so put the error message in.
      this.promise.reject("invoke method error", err?.message, null)
    }

    this.remove() // cleanup the promise
  }

  private fun store() {
    promises.add(this)
  }

  private fun remove() {
    promises.remove(this)
  }
}
